### ISSUE TITLE
Legion Prof: change provenance format

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1554]
My understanding is that current provenance strings take the form:

```
foo.py:42$OpCode,MULTIPLY|Provenance,foo.py:42
```

(from https://github.com/StanfordLegion/prof-viewer/pull/26/commits/e921afe001e844c9692c5b0834ccc83ac89a8128)

Where the part before `$` is a user-supplied string (just taken as is), and the second part is a machine-readable key-value map separated by `,` and `|`.

There are a couple problems with this:

 1. As far as I know, we have not defined an escape syntax. Therefore, you cannot include `$` in the user-defined portion or `,` or `|` in the machine-readable portion.
 2. Even if we do define an escape syntax, it's not realistic to expect users to follow it. This is a classic area where users mess up, leading to corrupted formatting that cannot be parsed.
 3. This is all a custom syntax and doesn't build on standard ways to format structured data.

I would like to propose an alternative. In this scheme, a provenance can be one of two things:

 1. user-readable (we do not attempt to parse it at all), OR
 2. machine-readable

It is one or the other. Not both.

My proposal is that machine readable strings always start with `{` and are JSON-formatted objects. Any value that does not start with `{` is a user-readable string and is not parsed. While this is technically a special character, I think it's very unlikely that a user-readable string will start with `{`. (If for some reason the user *really* wanted it, they could put another character, like a space, first.) There is no separator character that has to be escaped and no way to mess up that formatting.

When the string is machine readable, we can still embed a user-readable portion like so:

```
{"user": "The value of the user-readable string", ...}
```

In this way, there's no question about the formatting, it's just the key `user` holds the user-supplied string (if it exists). This removes the need to escape the user-supplied string.

Because JSON is a very common format, we can expect users to serialize it from any language. It is very well understood and provides broader capabilities than what we have now. This also avoids needing to write a custom parser in the profiler backend, as well as the need to define escape syntax to preserve arbitrary values in the output.

For posterity, the original provenance string in my format would be:

```json
{"user": "foo.py:42", "OpCode": "MULTIPLY", "Provenance": "foo.py:42"}
```

(Though in the new format `Provenance` is redundant and we can get rid of that, since it's included under `user`. But the format doesn't prohibit it; users can do what they want.)

CC @manopapad @eddy16112 for comments.

**Comment by manopapad** [comment]
I agree with all the above. For some context, here is a sketch for how we'd like this to work end-to-end in Legate (CC @magnatelee to fact-check me):

```
# cunumeric
def binary_op(op: string, rhs1: ndarray, rhs2: ndarray):
  with Annotation("Provenance", backtrace_of_parent().top()):
    with Annotation("Backtrace", backtrace_of_parent(), user_readable=False):
      with Annotation("OpCode", op, user_readable=False):
        # perform operation

# user code, foo.py
for i in range(iters):
  with Annotation("Iteration", i):
    x = x * y
```

The above should have the effect of reporting the following user-level provenance on Legion error messages:

```
LEGION ERROR: Task BinaryOp from foo.py:42[Iteration=0] took read-only privileges on uninitialized data
```

And show the following information on the profiler:

```
Provenance: foo.py:42[Iteration=0]
Backtrace: ...
OpCode: ADD
```

So I think in your scheme, what we would pass to the profiler would be:

```
{"user": "foo.py:42[Iteration=0]", "Backtrace": "<backtrace with newlines>", "OpCode": "ADD"}
```

**Comment by lightsighter** [comment]
This is pass-through from the runtime's perspective (Legion doesn't attempt to parse the string at all currently, but when it does it will just care about the `provenance` value) and is a contract between users and the profiler so I'm ambivalent. 😇 Seems overall reasonable to me though.

**Comment by magnatelee** [comment]
@elliottslaughter I also agree with all your points, but I vaguely remember the reason that we didn't go down that route in the first shot is that the runtime would want to easily extract the human readable part and use it in various places. (see below)

> This is pass-through from the runtime's perspective (Legion doesn't attempt to parse the string at all currently, but when it does it will just care about the provenance value) and is a contract between users and the profiler so I'm ambivalent. 😇 Seems overall reasonable to me though.

@lightsighter that's not true: https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion/legion_ops.cc?ref_type=heads#L57-67 The runtime has been using the human-readable portion for error reporting, and I feel it won't be ergonomic if the runtime embeds the whole JSON string in its error messages.

I see there are three options to handle machine readable JSONs:

1. Don't use provenance strings in error messages if they are machine readable
2. Add a JSON parser to Legion so it can extract the human readable portion from a JSON
3. Make provenance strings truly pass-through, whether they are human readable or not

I'm fine with any of those.

**Comment by elliottslaughter** [comment]
There is another option. We could require that machine-specified provenance always start with `{"user": "`, in which case you can just match that literal and then parse until the matching `"`. We'd need to handle escape syntax, but we'll need escapes anyway.

Alternatively, there may exist JSON parsers that basically give us the ability to do that: run a callback on each key (without allocating), and then when we match copy the decoded string and terminate early without parsing the rest of the JSON object.

At this point, to unblock this issue I think we need someone to do a little exploration on the available JSON options for C/C++ to see what is likely to be the best fit.

**Comment by lightsighter** [comment]
> At this point, to unblock this issue I think we need someone to do a little exploration on the available JSON options for C/C++ to see what is likely to be the best fit.

I think it needs to be someone who is planning to use the new syntax to know what is required.

**Comment by bryevdv** [comment]
> We could require that machine-specified provenance always start with {"user": ", in which case you can just match that literal and then parse until the matching "

JSON objects are defined as unordered collections of key/value pairs so we should not expect order to be preserved in case of any intermediate serialization, so I'd be :-1:  on this. 

@manopapad and I had discussed this awhile back and wondered about using an array instead. e.g. for

```python
for i in range(...):
  with Provenance(iter=):
     # do stuff
```
generate a "tuple" of *[user-part, machine-part]*:

```json
["test.py:42[iter=7]", {"file": "test.py", "line": 42, "iter": 7, "backtrace": "..."}]
```
Or in cases where there is no *user-part*, then
```json
[{"file": "..."}]
```
those cases seem distinguishable in case its desirable to avoid a JSON parser dependency.

**Comment by elliottslaughter** [comment]
I think it's fair to use an array as a tuple as long as we're clear that the length is fixed and we expect certain items in certain positions.

I'm not sure why we bother with the array in this case:

```
[{"file": "..."}]
```

I mean, if the tuple length is going to vary anyway, the format is not truly homogenou, so why not switch on kind and not on length?

Since we're going to have two cases anyway, we could do:

 1. A 2-tuple with the first item always being a string *user-part*:

   ```
   ["test.py:42[iter=7]", {"file": "test.py", "line": 42, "iter": 7, "backtrace": "..."}]
   ```

 2. An object (no *user-part*):

   ```
   {"file": "test.py", "line": 42, "iter": 7, "backtrace": "..."}
   ```

**Comment by bryevdv** [comment]
Sure, that seems fine too.

**Comment by elliottslaughter** [comment]
While the JSON spec is not really that complicated (it can fit on a single page: https://www.json.org/json-en.html), I do think it's worth at least parsing backslash-escapes. Otherwise it will be impossible for the user to embed literal `"` characters (and just generally defy what users expect of JSON).

Writing a simplistic parser that looks for literally `["` and then decodes the string shouldn't be too bad, but does still require looking at characters one at a time to find `\` in the middle of the string (which then need to be transcoded). A naive implementation might be slow.

Otherwise, if we have a JSON decoder that allow us to bail out early and does not force us to decode the entire object, that would work.

**Comment by lightsighter** [comment]
Just a detail: I don't want to add dependences for Legion on anything other than Realm. If we're going to parse JSON, then we're going to need to write our own parser for it. Not saying that is hard, but it is additional work that will need to be done by someone.

**Comment by manopapad** [comment]
How about vendoring a header-only JSON parsing library with a permissive license?

**Comment by lightsighter** [comment]
I mean if it is public-domain then I don't mind including it, but I don't want to have to have someone else's license in the Legion repo.

**Comment by elliottslaughter** [comment]
Ok, here you go. This parses the user-part string prefix and then leaves the rest alone:

```c++
#include <string>
#include <sstream>
#include <cassert>

bool parse_provenance_user_part(std::string &input, std::string &output) {
  if (input.length() <= 3 || input[0] != '[' || input[1] != '"') {
    return false;
  }
  std::stringstream ss;
  size_t len = input.length();
  bool closed = false;
  for (size_t i = 2; i < len; ++i) {
    if (input[i] == '\\') {
      if (i+1 >= len) {
        return false;
      }
      switch (input[i+1]) {
        case '"':
        case '\\':
        case '/':
          ss << input[i+1];
          break;
        case 'b':
          ss << '\b';
          break;
        case 'f':
          ss << '\f';
          break;
        case 'n':
          ss << '\n';
          break;
        case 'r':
          ss << '\r';
          break;
        case 't':
          ss << '\t';
          break;
        case 'u':
          return false; // Unicode is unsupported
        default:
          return false; // Bad escape
      }
      ++i;
    } else if (input[i] == '"') {
      closed = true;
      break;
    } else {
      ss << input[i];
    }
  }

  if (!closed) {
    return false;
  }

  output = ss.str();
  return true;
}

void test(const char *input, const char *output, bool expected) {
  std::string input_str(input);
  std::string output_str;

  bool ok = parse_provenance_user_part(input_str, output_str);
  assert(ok == expected);
  if (ok) {
    assert(output_str == std::string(output));
  }
}

int main() {
  test("[\"asdf\", {}]", "asdf", true);
  test("[\"Hello \\\"World\\\"!\", {}]", "Hello \"World\"!", true);


  // Partial prefixes do not parse unless they contain a complete string
  test("", NULL, false);
  test("[", NULL, false);
  test("[\"", NULL, false);
  test("[\"asdf", NULL, false);
  test("[\"asdf\"", "asdf", true);

  // Unrelated junk does not parse
  test("{", NULL, false);
  test("\"", NULL, false);
  test("1.0", NULL, false);
  test("null", NULL, false);

  // Bad escape
  test("[\"\\x\"", NULL, false);
  test("[\"\\u1234\"", NULL, false);
}
```

Run it for the test cases.

**Comment by bryevdv** [comment]
@elliottslaughter here is an updated version that extracts both parts
```c++
    //--------------------------------------------------------------------------
    void Provenance::initialize(const char *prov, size_t size)
    //--------------------------------------------------------------------------
    {
      if (prov[0] == '{') {
        machine.assign(prov);
      }
      else {
        std::string input{prov};
        parse_provenance_parts(input, human, machine);
      }
    }

    //--------------------------------------------------------------------------
    /*static*/ bool Provenance::parse_provenance_parts(std::string &input,
                             std::string &human_part, std::string &machine_part)
    //--------------------------------------------------------------------------
    {
      size_t len = input.length();

      // shortest valid input: ["",{}]
      if (input.length() < 7) {
        return false;
      }

      // must start with: ["
      if (input[0] != '[' || input[1] != '"') {
        return false;
      }

      // must end with: }]
      if (input[len-2] != '}' || input[len-1] != ']') {
        return false;
      }


      std::stringstream ss_h;
      bool human_closed = false;

      size_t i = 2;
      for (; i < len; ++i) {
        if (input[i] == '\\') {;
          if (i+1 >= len) {
            return false;
          }
          switch (input[i+1]) {
            case '"':
            case '\\':
            case '/':
              ss_h << input[i+1];
              break;
            case 'b':
              ss_h << '\b';
              break;
            case 'f':
              ss_h << '\f';
              break;
            case 'n':
              ss_h << '\n';
              break;
            case 'r':
              ss_h << '\r';
              break;
            case 't':
              ss_h << '\t';
              break;
            case 'u':
              return false; // Unicode is unsupported
            default:
              return false; // Bad escape
          }
          ++i;
        } else if (input[i] == '"') {
          human_closed = true;
          break;
        } else {
          ss_h << input[i];
        }
      }

      if (!human_closed) {
        return false;
      }

      human_part = ss_h.str();

      for (; i < len-1; ++i) {
        if (input[i] == '{') {
          break;
        }
      }

      // machine part never opened
      if (i == len-1) {
        return false;
      }

      machine_part.assign(input.c_str() + i, input.c_str() + len - 1);

      return true;
    }
```

And expanded tests:

<details>

```c++
void test(const char *input, const char *h, const char* m, bool expected) {
  std::string input_str(input);
  std::string h_str;
  std::string m_str;

  bool ok = parse_provenance_parts(input_str, h_str, m_str);
  assert(ok == expected);
  if (ok) {
    assert(h_str == std::string(h));
    assert(m_str == std::string(m));
  }
}

int main() {
  test("[\"\",{}]", "", "{}", true);
  test("[\"\"   ,  {}]", "", "{}", true);

  test("[\"asdf\", {}]", "asdf", "{}", true);
  test("[\"Hello \\\"World\\\"!\", {}]", "Hello \"World\"!", "{}", true);

  test("[\"\", {\"foo\": 10}]", "", "{\"foo\": 10}", true);
  test("[\"asdf\", {\"foo\": 10}]", "asdf", "{\"foo\": 10}", true);

  // Partial prefixes do not parse
  test("", NULL, NULL, false);
  test("[", NULL, NULL, false);
  test("[\"", NULL, NULL, false);
  test("[\"asdf", NULL, NULL, false);
  test("[\"asdf\"", "asdf", NULL, false);
  test("[\"asdf\"]", "asdf", NULL, false);
  test("[\"asdf\",]", "asdf", NULL, false);
  test("[\"asdf\",{]", "asdf", NULL, false);
  test("[\"asdf\",{\"foo\": 10]", "asdf", NULL, false);

  // Unrelated junk does not parse
  test("{", NULL, NULL, false);
  test("\"", NULL, NULL, false);
  test("1.0", NULL, NULL, false);
  test("null", NULL, NULL, false);

  // Bad escape
  test("[\"\\x\"", NULL, NULL, false);
  test("[\"\\u1234\"", NULL, NULL, false);
}
```

</details>


I was going top open a PR but had some questions/issues:

* I don't see any existing tests for `Provenance` under `tests` Are they unnecessary? 
* What do you want to happen in case `parse_provenance_parts` fails?
* There seems to be ton of trailing blank space lines that my editor cleans up by default, are code-gardening changes acceptable?
* Is there a formatter run on this codebase? 

I did have a look in the README but did not see answers

**Comment by elliottslaughter** [comment]
I think these are questions for @lightsighter.

**Comment by manopapad** [comment]
^ ping @lightsighter on the above questions

**Comment by lightsighter** [comment]
> I don't see any existing tests for Provenance under tests Are they unnecessary?

If you want to add unit tests for provenance to the `test` subdirectory you can, but I won't insist on it.

> What do you want to happen in case parse_provenance_parts fails?

What failure modes do you envision? If we can't find the delimiter, I expect we're just going to assume the whole thing is the user part of the string and WYSIWYG.

> There seems to be ton of trailing blank space lines that my editor cleans up by default, are code-gardening changes acceptable?

No, they will cause conflicts with other major branches that are outstanding.

> Is there a formatter run on this codebase?

Not until later this year. Do whatever you want as long as lines are less than 80 characters and you don't use the `auto` keyword.

**Comment by bryevdv** [comment]
@lightsighter @manopapad https://gitlab.com/StanfordLegion/legion/-/merge_requests/1384

**Comment by lightsighter** [comment]
Localized merge request: https://gitlab.com/StanfordLegion/legion/-/merge_requests/1392

**Comment by elliottslaughter** [comment]
Update from the meeting today:

We are planning to merge https://gitlab.com/StanfordLegion/legion/-/merge_requests/1392 once @lightsighter gets a chance to review it.

Note this change only makes the runtime parse the user portion of the provenance string.

Rust Legion Prof does NOT attempt to parse provenance strings (whether in old or new formats), and thus we have NO test coverage of either new or old formats. (Note that up to about ~2 weeks ago, Rust and Python were tested in CI to produce bitwise identical output in all cases. Therefore there cannot possibly have been any testing, as Rust has never implemented this.) There may be existing Python support but if so it has never been tested in our CI.

Someone please provide me with an example of a profile that generates the format you want, so that I can add the appropriate Rust-side support for this. @manopapad @magnatelee

**Comment by lightsighter** [comment]
These changes are merge into Legion. Do we want to keep the issue open until we fix the profiler or start a new issue?

**Comment by manopapad** [comment]
@bryevdv could you please write a pure-Legion-C++-API example that sets a provenance with user and machine-readable parts, so @elliottslaughter can write the profiler bits required to handle it?

**Comment by bryevdv** [comment]
@manopapad @elliottslaughter Here is a diff from `examples/provenance` I think this is what you are asking for
```diff
diff --git a/examples/provenance/provenance.cc b/examples/provenance/provenance.cc
index 622136103..39a3cfd01 100644
--- a/examples/provenance/provenance.cc
+++ b/examples/provenance/provenance.cc
@@ -91,23 +91,39 @@ void top_level_task(const Task *task,
   // IndexSpace is = runtime->create_index_space(ctx, elem_rect, 0, "Element IndexSpace");
   Domain elem_domain = Domain(elem_rect);
   Future bound_future = Future::from_value<Domain>(elem_domain);
-  std::string is_prov = "Element IndexSpace:" + std::to_string(__LINE__);
+
+  // human part + empty machine part ["stuff", {}]
+  std::string is_prov = "[\"Element IndexSpace:" + std::to_string(__LINE__) + "\", {}]";
+  printf("PROV: %s\n", is_prov.c_str());
+
   IndexSpace is = runtime->create_index_space(ctx, 1, bound_future, 0, is_prov.c_str()); 
   runtime->attach_name(is, "is");
   Future field_size_future = Future::from_value<size_t>(sizeof(double));
   std::vector<Future> field_sizes{field_size_future, field_size_future};
   std::vector<FieldID> field_ids{FID_X, FID_Y};
-  std::string field_xy_prov = "Element FieldSpace XY:" + std::to_string(__LINE__);
+
+  // empty human part + machine part ["", {"key": "stuff"}]
+  std::string field_xy_prov = "[\"\", {\"key\": \"Element FieldSpace XY:" + std::to_string(__LINE__) + "\"}]";
+  printf("PROV: %s\n", field_xy_prov.c_str());
+
   FieldSpace fs = runtime->create_field_space(ctx, field_sizes, field_ids, 0, field_xy_prov.c_str());
   runtime->attach_name(fs, "fs");
   {
     FieldAllocator allocator = 
       runtime->create_field_allocator(ctx, fs);
+
+    // human part (string) only
     std::string field_z_prov = "Element FieldSpace Z:" + std::to_string(__LINE__);
+    printf("PROV: %s\n", field_z_prov.c_str());
+
     allocator.allocate_field(sizeof(double),FID_Z, 0, false, field_z_prov.c_str());
     runtime->attach_name(fs, FID_Z, "Z");
   }
-  std::string input_lr_prov = "Input LR:" + std::to_string(__LINE__);
+
+  // machine part (object) only
+  std::string input_lr_prov = "{\"key\": \"Input LR:" + std::to_string(__LINE__) + "\"}]";
+  printf("PROV: %s\n", input_lr_prov.c_str());
+
   LogicalRegion input_lr = runtime->create_logical_region(ctx, is, fs, false, input_lr_prov.c_str());
   runtime->attach_name(input_lr, "input_lr");
   std::string output_lr_prov = "Output LR:" + std::to_string(__LINE__);
```

Running uses these provenance string in various places
```
PROV: ["Element IndexSpace:96", {}]
PROV: ["", {"key": "Element FieldSpace XY:106"}]
PROV: Element FieldSpace Z:116
PROV: {"key": "Input LR:124"}]
```

**Comment by manopapad** [comment]
@elliottslaughter did you get a chance to add the handling of the new provenance format to the rust profiler?

**Comment by elliottslaughter** [comment]
In progress MR at: https://gitlab.com/StanfordLegion/legion/-/merge_requests/1460

**Comment by elliottslaughter** [comment]
Can someone confirm that the patch in https://github.com/StanfordLegion/legion/issues/1554#issuecomment-2259349371 is doing what you want? If I'm not mistaken, the provenance is being attached to index spaces, fields and regions. I don't see any provenance being attached to tasks, and therefore there is nothing to see in the profile?

**Comment by manopapad** [comment]
There is provenance being attached to task launchers, but it's only a user part (a single string, no machine part). Maybe change one of them in the example to add a machine part as well?

**Comment by elliottslaughter** [comment]
If you want a meaningful test of the new functionality, you will need to modify at least one task, copy or instance.

**Comment by elliottslaughter** [comment]
I updated the example to provide a nontrivial example with the DAXPY task. Here's what it looks like:

<img width="597" alt="Screenshot 2024-09-10 at 9 56 16 AM" src="https://github.com/user-attachments/assets/ac670267-0080-4999-96a8-ce622092a478">

Are you ok with the quotes around the values, or would you prefer for me to remove them if possible? @manopapad

**Comment by manopapad** [comment]
We can live with the quotes (assuming they're not trivial to remove, otherwise you wouldn't be asking us about it).

Do you want me/Bryan to update the example on the PR, to add machine provenance to other relevant operations, so it's tested?

**Comment by elliottslaughter** [comment]
Actually, removing the quotes is pretty easy, like another 3 lines or so. The main thing to understand is that it will work *only* for JSON strings. So if you provenance is:

```
["", {"key": {"some": "object"}}]
```

Then it will show up in the display as `key: {"some": "object"}`, and I think there is no way around that without defining the semantic content of complex (non-string) provenance values.

**Comment by elliottslaughter** [comment]
> Do you want me/Bryan to update the example on the PR, to add machine provenance to other relevant operations, so it's tested?

Yes, please see the current contents of https://gitlab.com/StanfordLegion/legion/-/merge_requests/1460 and send me a patch if there is anything else you want me to test.

**Comment by manopapad** [comment]
AFAIK, we intend to pass numbers and strings only to the machine provenance, e.g. `{"file": "foo.py", "line": 42}`, not structured objects. So it would be slightly better for our use to skip the quotes, especially so that a full backtrace (if we ever pass something like that) would be nicely formatted like

```
          bar.py:12
backtrace foo.py:14
          qux.py:33
```

instead of

```
          "bar.py:12
backtrace foo.py:14
          qux.py:33"
```

As for a patch, I guess it's good to leave some usecases w/o machine provenance, so we test the "pure user provenance" part as well. One thing I'd suggest is to test with a pure number for the line:

```
diff --git a/examples/provenance/provenance.cc b/examples/provenance/provenance.cc
index d2116ef84..8b2aa20c3 100644
--- a/examples/provenance/provenance.cc
+++ b/examples/provenance/provenance.cc
@@ -218,7 +218,7 @@ void top_level_task(const Task *task,
   const double alpha = 0.1;
   IndexLauncher daxpy_launcher(DAXPY_TASK_ID, color_is,
                 TaskArgument(&alpha, sizeof(alpha)), arg_map);
-  std::string daxpy_launcher_prov = "[\"Daxpy IndexLauncher:" + std::to_string(__LINE__) + "\", {\"task\": \"Daxpy\", \"launcher\": \"IndexLauncher\", \"line\": \"" + std::to_string(__LINE__) + "\"}]";
+  std::string daxpy_launcher_prov = "[\"Daxpy IndexLauncher:" + std::to_string(__LINE__) + "\", {\"task\": \"Daxpy\", \"launcher\": \"IndexLauncher\", \"line\": " + std::to_string(__LINE__) + "}]";
   daxpy_launcher.provenance = daxpy_launcher_prov;
   daxpy_launcher.add_region_requirement(
       RegionRequirement(input_lp, 0/*projection ID*/,
```

**Comment by elliottslaughter** [comment]
MR 1460 has been merged, please go ahead and close this issue when you're ready.
